## This is the implementation of Dataquest's "Exploring eBay Car Sales Data" guided project, which, as the name suggests, uses a dataset of eBay car listings in Germany.

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Observations: The dataset seems to be mostly intact. Most of the columns contain the full 50K entries, but a few of them such as "notRepairedDamage" and "vehicleType" are missing quite a few thousand entries.

## Cleaning Column Names

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In this section we basically just cleaned up the name of the columns without altering any data.

## Data Exploration

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Apparently the columns "seller", "offer_type", "abtest", "vehicle_type", "gearbox" and "odometer" all have very low numbers of unique entries, which may indicate they are good candidates for being dropped. 

Strangely, "registration_year", "registration_month", "nr_of_pictures" and "postal_code" seem to return NAN, which is counter intuitive.

Let's investigate this further.

In [8]:
autos.registration_year.value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

In [9]:
autos.nr_of_pictures.value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

In [10]:
autos.postal_code.value_counts()

10115    109
65428    104
66333     54
45888     50
44145     48
        ... 
23942      1
83365      1
95683      1
97794      1
67585      1
Name: postal_code, Length: 7014, dtype: int64

In [11]:
autos.registration_month.value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

Analyzing theses results we can tell that, appart from "nr_of_pictures", all other columns with NAN in unique actually have many distinct values. For some reason describe() isn't being able to parse this data correctly. Therefore, we are going to drop "nr_of_pictures".

In [12]:
autos = autos.drop(["nr_of_pictures"], axis=1)

In [13]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00        79588   
1         bmw              nein  2016-04-04 00:00:00        71034   
2  volkswagen              nein  2016-03-26 00:00:00        35394   
3       smart              nein  2016-03-12 00:00:00        33729   
4        ford              nein  2016-04-01 00:00:00        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [14]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                 object
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer              object
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
postal_code            int64
last_seen             object
dtype: object

From the code above we can tell which columns are objects (in this case strings) and which are integer numbers. We wish to change "price" and "odometer" from strings to numbers.

In [15]:
autos.price = (autos.price
                    .str.replace("$","")
                    .str.replace(",","")
                    .astype(int))
autos.price.head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [16]:
autos.odometer = (autos.odometer
                    .str.replace("km","")
                    .str.replace(",","")
                    .astype(int))
autos.rename({"odometer" : "odometer_km"}, axis=1, inplace=True)
autos.odometer_km.head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## Continuing Exploration of Columns

We will continue exploring odometer_km and price to try and find outliers so we can remove them.

In [17]:
autos.odometer_km.unique().shape

(13,)

In [18]:
autos.price.unique().shape

(2357,)

In [19]:
autos.odometer_km.describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos.price.describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [21]:
autos.odometer_km.value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

There doesn't seem to be anything unusual with the ranges of the odometer_km column, so we will leave it as is.

In [22]:
autos.price.value_counts().sort_index(ascending=True)

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

It seems there are a few cars both on the very low end (zero euros) and very high end (99 million euros) of prices. To fix this we could create a minimum and maximum value and then filter our price column to these values. I arbitrarily chose that 1 euro and 1 million euros would be the minimum and maximum, respectively.

In [23]:
autos = autos[autos.price.between(1, 1000000)]

In [24]:
autos.price.value_counts().sort_index(ascending=True)

1         156
2           3
3           1
5           2
8           1
         ... 
299000      1
345000      1
350000      1
999990      1
999999      2
Name: price, Length: 2348, dtype: int64

## Understanding Date Ranges

In [25]:
(autos.date_crawled
             .str[:10]
             .value_counts(normalize=True, dropna=False)
             .sort_index())

2016-03-05    0.025325
2016-03-06    0.014042
2016-03-07    0.036011
2016-03-08    0.033294
2016-03-09    0.033088
2016-03-10    0.032182
2016-03-11    0.032573
2016-03-12    0.036917
2016-03-13    0.015669
2016-03-14    0.036547
2016-03-15    0.034282
2016-03-16    0.029608
2016-03-17    0.031646
2016-03-18    0.012910
2016-03-19    0.034776
2016-03-20    0.037885
2016-03-21    0.037391
2016-03-22    0.032985
2016-03-23    0.032223
2016-03-24    0.029340
2016-03-25    0.031605
2016-03-26    0.032202
2016-03-27    0.031090
2016-03-28    0.034858
2016-03-29    0.034117
2016-03-30    0.033685
2016-03-31    0.031832
2016-04-01    0.033685
2016-04-02    0.035476
2016-04-03    0.038606
2016-04-04    0.036485
2016-04-05    0.013095
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

From exploring date_crawled we can gather that the crawling was made over roughly a month, from March 5th, 2016 to April 7th, 2016. Beyond that, there doesn't seem to be any day which had a lot more crawling done than any other.

In [26]:
(autos.last_seen
             .str[:10]
             .value_counts(normalize=True, dropna=False)
             .sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005394
2016-03-08    0.007412
2016-03-09    0.009595
2016-03-10    0.010665
2016-03-11    0.012374
2016-03-12    0.023781
2016-03-13    0.008895
2016-03-14    0.012601
2016-03-15    0.015875
2016-03-16    0.016451
2016-03-17    0.028084
2016-03-18    0.007351
2016-03-19    0.015833
2016-03-20    0.020651
2016-03-21    0.020631
2016-03-22    0.021372
2016-03-23    0.018531
2016-03-24    0.019766
2016-03-25    0.019210
2016-03-26    0.016801
2016-03-27    0.015648
2016-03-28    0.020878
2016-03-29    0.022360
2016-03-30    0.024769
2016-03-31    0.023781
2016-04-01    0.022793
2016-04-02    0.024914
2016-04-03    0.025202
2016-04-04    0.024481
2016-04-05    0.124753
2016-04-06    0.221813
2016-04-07    0.131939
Name: last_seen, dtype: float64

The last_seen column is pretty similar to date_crawled, even starting and ending on the same days.

In [27]:
(autos.ad_created
             .str[:10]
             .value_counts(normalize=True, dropna=False)
             .sort_index())

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038853
2016-04-04    0.036856
2016-04-05    0.011818
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

ad_created is a little bit different from the two before, since it has some older listings, going back as far as the 11th of june, 2015.

Now, exploring the distribution of registration_year:

In [28]:
autos.registration_year.describe()

count    48568.000000
mean      2004.754612
std         88.641262
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

From this result we can straight up tell that some entries are wrong, since the minimum year is 1000, and the maximum is 9999, which shows we need to clean up this column's data. Let's find out how many of our entries lie beyond 1900-2016:

In [29]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038790973480480974

Since only 3.8% of our data lie beyond these boundaries, let's remove these outliers:

In [30]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [32]:
autos["registration_year"].value_counts(normalize=True)

2000    0.067603
2005    0.062891
1999    0.062056
2004    0.057900
2003    0.057814
          ...   
1938    0.000021
1948    0.000021
1927    0.000021
1931    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

From the percentage of each year, aquired above, it seems most cars were registere in the past 30 or so years.

## Exploring Brands

In [34]:
autos.brand.value_counts()

volkswagen        9864
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3264
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

We can see that the top 5 most popular car brands are german, which is kind of fitting since theses listings are from German eBay. Anyway, we will compare the average pricing from the top 5 brands to the bottom 5 brands:

In [35]:
top_5_brands = ['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi']

In [36]:
top_5_mean_prices = {}
for brand in top_5_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    top_5_mean_prices[brand] = int(mean_price)

top_5_mean_prices

{'volkswagen': 5604,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336}

In [37]:
bottom_5_brands = ['lada', 'lancia', 'rover', 'trabant', 'daewoo']

In [38]:
bottom_5_mean_prices = {}
for brand in bottom_5_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    bottom_5_mean_prices[brand] = int(mean_price)

bottom_5_mean_prices

{'lada': 2688, 'lancia': 3376, 'rover': 1602, 'trabant': 1790, 'daewoo': 1049}

Comparing the top brands and bottom brands we can infer that the consumer preference doesn't come from better pricing, since almost all top 5 brands (except OPEL) have higher average prices than the bottom 5. Therefore, the german market preference might come from product quality or other less objective reasons such as taste and culture.

## Exploring Mean Mileage from Brands

In [41]:
top_5_mean_mileage = {}
for brand in top_5_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["odometer_km"].mean()
    top_5_mean_mileage[brand] = int(mean_price)

top_5_mean_mileage

{'volkswagen': 128711,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157}

In [42]:
mean_mileage = pd.Series(top_5_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(top_5_mean_prices).sort_values(ascending=False)

In [47]:
brand_info = pd.DataFrame(mean_mileage, columns=['mean_mileage'])
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128711        5604

There doesn't seem to be any relation between average mileage and average car price from what we can gather in this table. Overall, the conclusion we could achieve is that BMWs, Mercedes and Audis are just more expensive cars.